File: Lesson 4 - Counting Elements/PermCheck.cpp
#include <iostream>
#include <assert.h>

#include <algorithm>
#include <limits.h>
#include <vector>

int solution(std::vector<int> &A)
{
    std::vector<bool> isPresent(A.size(), false);

    const int64_t size = static_cast<int64_t>(A.size());
    for (const auto& num : A)
    {
        if (num > size)
        {
            return false;
        }

        isPresent[num - 1] = true;
    }

    return std::all_of(isPresent.begin(), isPresent.end(), [](const bool& val) { return val; });
}

int main()
{
    assert(1 == solution(std::vector<int>()));
    assert(1 == solution(std::vector<int>({ 1 })));
    assert(0 == solution(std::vector<int>({ 2 })));
    assert(1 == solution(std::vector<int>({ 4, 1, 3, 2 })));
    assert(0 == solution(std::vector<int>({ 4, 1, 3 })));

    const int MaxArrSize = 100000; // by the task
    std::vector<int> stressTestArr(MaxArrSize);
    for (size_t i = 0; i < stressTestArr.size(); i++)
    {
        stressTestArr[i] = i + 1;
    }
    assert(1 == solution(stressTestArr));

    stressTestArr.back() = INT_MAX;
    assert(0 == solution(stressTestArr));
}

File: Lesson 4 - Counting Elements/PermCheck.py
from sys import maxint

def solution(A):
    isPresent = [0] * len(A)

    for num in A:
        if num > len(A):
            return 0

        isPresent[num - 1] = 1

    return int(sum(isPresent) == len(isPresent))


assert 1 == solution([]);
assert 1 == solution([ 1 ]);
assert 0 == solution([ 2 ]);
assert 1 == solution([ 4, 1, 3, 2 ]);
assert 0 == solution([ 4, 1, 3 ]);

stressTestArr = range(100000) # by the task
assert 1 == solution(stressTestArr)
stressTestArr[-1] = maxint
assert 0 == solution(stressTestArr)
